internal/rotate.go

```diff
@@ -36,120 +25,56 @@ type CertRotationState struct {
 }
 
 type CertRotationHandler struct {
-	State     CertRotationState
-	stateFile string
-	app       *App
-	client    *http.Client
-	crypto    *EciesCrypto
-	eventSync EventSync
-	steps     []CertRotationStep
-	cienv     bool
+	stateHandler
+	State CertRotationState
 }
 
-type CertRotationStep interface {
+type certRotationStep interface {
 	Name() string
 	Execute(handler *CertRotationHandler) error
 }
 
-// NewCertRotationHandler constructs a new handler to initiate a rotation with
 func NewCertRotationHandler(app *App, stateFile, estServer string) *CertRotationHandler {
-	eventUrl := app.sota.GetOrDie("tls.server") + "/events"
+	handler := &CertRotationHandler{
+		State:        CertRotationState{EstServer: estServer},
+		stateHandler: newStateHandler(app, stateFile),
+	}
 
-	target, err := LoadCurrentTarget(filepath.Join(app.StorageDir, "current-target"))
-	if err != nil {
-		log.Printf("Unable to parse current-target. Events posted to server will be missing content: %s", err)
+	adapter := func(step certRotationStep) stateStep {
+		return stateStep{
+			Name: step.Name(),
+			Execute: func(h *stateHandler) error {
+				// Golang does not allow to convert stateHandler to CertRotationHandler
+				// So, use closure to hold the reference to it.
+				return step.Execute(handler)
+			},
+		}
 	}
```

Comment:
This is a narrow place here.

Due to the Golang's type model limitations, there is no way to cast stateHandler to CertRotationHandler here.

There are basically two ways how this could be achieved:
1. Via generics this could be done seamlessly with some magic. See the other PR for this.
2. Via interfaces there could be a classic interface case here. But, using interfaces would result into too much code rewrite, and we'd still need either this adapter, or a cast inside each of the step types.